contrib/raftexample/raft.go

```diff
@@ -287,17 +428,45 @@ func (rc *raftNode) serveChannels() {
 		case <-ticker.C:
 			rc.node.Tick()
 
-		// store raft entries to wal, then publish over commit channel
+			// store raft entries to wal, then publish over commit channel
 		case rd := <-rc.node.Ready():
-			rc.wal.Save(rd.HardState, rd.Entries)
+			isLeader := false
+			if rd.SoftState != nil && rd.SoftState.RaftState == raft.StateLeader {
+				isLeader = true
+			}
+
+			applyDone := make(chan bool)
+			go rc.applyAll(pr, &apply{
+				entriesToApply: rd.CommittedEntries,
+				snapshotToSave: rd.Snapshot,
+				applyDone:      applyDone,
+			})
+
+			// (Raft §10.2.1 Writing to the leader’s disk in parallel, p.141)
+			// leader writes the new log entry to disk before replicating the entry
+			// to its followers. Then, the followers write the entry to their disks.
+			// Fortunately, the leader can write to its disk in parallel with replicating
+			// to the followers and them writing to their disks.
+			if isLeader {
+				rc.transport.Send(rd.Messages)
+			}
+
+			rc.storage.Save(rd.HardState, rd.Entries)
+			if !raft.IsEmptySnap(rd.Snapshot) {
+				rc.storage.SaveSnap(rd.Snapshot)
+				rc.raftStorage.ApplySnapshot(rd.Snapshot)
+			}
 			rc.raftStorage.Append(rd.Entries)
-			rc.transport.Send(rd.Messages)
-			if ok := rc.publishEntries(rd.CommittedEntries); !ok {
```

Comment:
it'd be much simpler to keep the publish bit synchronous instead of overlapping applies ala etcdserver